### PR TITLE
Hearing worksheet downloader link

### PR DESCRIPTION
**Pull request by nolawi**
connect #3601
screen shot 2017-12-14 at 4 19 41 pm

To test:
- Load welcome gate, click on hearing worksheet, confirm it loads correctly
- Confirm reader users cannot access other hearing prep pages
- Confirm hearing prep users can still access print page

### CODE REVIEW

File: client/app/reader/utils.js [hunk 1]
```diff
@@ -128,16 +129,22 @@ export const isUserEditingText = () => _.some(
...
+  let worksheetPdfLink;
+  let appealHasHearing = claim.hearing_id;
 
   if (claim.cavc && claim.aod) {
     appealTypeInfo = 'AOD, CAVC';
   } else if (claim.cavc) {
     appealTypeInfo = 'CAVC';
   } else if (claim.aod) {
     appealTypeInfo = 'AOD';
+    worksheetPdfLink = <a target="_blank" href={`/hearings/${appealHasHearing}/worksheet`}>Hearing Worksheet</a>;
```
Comment (sharonwarner) [inline client/app/reader/utils.js]: We want to send the attorney to the print page, but the editable hearing worksheet.

File: client/app/reader/utils.js [hunk 2]
```diff
@@ -128,16 +129,22 @@ export const isUserEditingText = () => _.some(
...
   }
 
-  return <span className="cf-red-text">{appealTypeInfo}</span>;
+  return <div className="claim-detail-container"><span className="claim-detail-type-info">{appealTypeInfo}</span>
+    {appealHasHearing && <span>{worksheetPdfLink}<DownloaderIcon className="downloader" /></span> }
+  </div>;
```
Comment (sharonwarner) [inline client/app/reader/utils.js]: Sorry for the confusion - this is my fault. We want to display the hearing worksheet for any case that has a worksheet in our database, not just AOD cases.

File: app/models/appeal.rb [hunk 3]
```diff
@@ -454,14 +454,15 @@ def latest_event_date
...
-  def to_hash(viewed: nil, issues: nil)
+  def to_hash(viewed: nil, issues: nil, hearing_id: nil)
     serializable_hash(
       methods: [:veteran_full_name, :docket_number, :type, :cavc, :aod],
       includes: [:vbms_id, :vacols_id]
     ).tap do |hash|
       hash["viewed"] = viewed
       hash["issues"] = issues
       hash["regional_office"] = regional_office_hash
+      hash["hearing_id"] = hearing_id
```
Comment (sharonwarner) [inline app/models/appeal.rb]: 👍

File: client/app/reader/utils.js (hunk 2 above)
Comment (nolawi) [reply client/app/reader/utils.js]: ok.. all cases...

File: app/models/user.rb [hunk 4]
```diff
@@ -101,7 +101,8 @@ def current_case_assignments_with_views
     opened_appeals = viewed_appeals(appeals.map(&:id))
 
     appeals.map do |appeal|
-      appeal.to_hash(viewed: opened_appeals[appeal.id], issues: appeal.issues)
+      hearing_id = !appeal.hearings.empty? ? appeal.hearings[0].id : nil
```
Comment (sharonwarner) [inline app/models/user.rb]: This is going to call VACOLS to get the hearing id. We want to just get it from our database. Something like Hearings.where(appeal_id: appeal.id) should work.

File: client/app/reader/utils.js [hunk 5]
```diff
@@ -128,6 +130,7 @@ export const isUserEditingText = () => _.some(
 
 export const getClaimTypeDetailInfo = (claim) => {
   let appealTypeInfo = '';
+  let appealHasHearing = 7;
```
Comment (sharonwarner) [inline client/app/reader/utils.js]: Was this an accidental commit?

Comment (nolawi) [reply client/app/reader/utils.js]: Yap - to Test

File: app/models/user.rb [hunk 6]
```diff
@@ -102,7 +102,8 @@ def current_case_assignments_with_views
     opened_appeals = viewed_appeals(appeals.map(&:id))
 
     appeals.map do |appeal|
-      appeal.to_hash(viewed: opened_appeals[appeal.id], issues: appeal.issues)
+      hearing = Hearing.find_by(appeal_id: appeal.id)
```
Comment (cmgiven) [inline app/models/user.rb]: Appeal->Hearing is 1->many. The use of `find_by` obscures that. How should the case of multiple hearings be handled?

**Comment by cmgiven** [comment]
The AC says to include the link only when a worksheet is present. If I'm understanding correctly, this would show empty worksheets.

**Comment by nolawi** [comment]
Aha we are doing more in the front end.. won't display anything if no hearing is present.. no hearing no worksheet.

**Comment by cmgiven** [comment]
Right, but just because there is a hearing does not mean there is a worksheet.

**Comment by sharonwarner** [comment]
I believe we fixed the API status bug where we were creating hearings accidentally. The hearings table stores most of the information from the worksheet, so I'm not sure of a better way to determine if a worksheet exists.

**Comment by mdbenjam** [comment]
How do I test this code? I don't see any test data changes?

File: client/app/reader/utils.js [hunk 7]
```diff
@@ -128,6 +130,7 @@ export const isUserEditingText = () => _.some(
 
 export const getClaimTypeDetailInfo = (claim) => {
   let appealTypeInfo = '';
+  let appealHasHearing = claim.hearing;
```
Comment (nolawi) [inline client/app/reader/utils.js]: here

File: client/app/components/RenderFunctions.jsx [hunk 8]
```diff
@@ -355,6 +355,22 @@ export class DownloadIcon extends React.PureComponent {
   }
 }
 
+export class DownloaderIcon extends React.PureComponent {
```
Comment (mdbenjam) [inline client/app/components/RenderFunctions.jsx]: Why add this? We already have the DownloadIcon component.

Comment (nolawi) [reply client/app/components/RenderFunctions.jsx]: its quite different- i couldn't re-use without modifying the way reader is using it..

Comment (mdbenjam) [reply client/app/components/RenderFunctions.jsx]: How is it different? In color? Or other things? If it's just color you can do something like this: https://github.com/department-of-veterans-affairs/caseflow/blob/issue3601/app/assets/stylesheets/_reviewer.scss#L346

File: app/models/user.rb (hunk 6 above)
Comment (sharonwarner) [reply app/models/user.rb]: Just confirmed with @lakohl, let's display all worksheets associated with the appeal.

**Comment by sharonwarner** [comment]
Good call. @nolawi let's give Reader users access to just the print version of the worksheet page.

File: client/app/components/RenderFunctions.jsx (hunk 8 above)
Comment (nolawi) [reply client/app/components/RenderFunctions.jsx]: ok _ I will make the change..

**Comment by gnakm** [comment]
**_If_** it's the print version, I think it's okay to have it in the case detail accordion. Originally we took it out of this section but I think it's safe to leave it there.

File: client/app/reader/utils.js [hunk 9]
```diff
@@ -126,8 +128,26 @@ export const isUserEditingText = () => _.some(
   (elem) => document.activeElement === elem
 );
 
+export const GetHearingWorksheetLink = (claim) => {
```
Comment (sharonwarner) [inline client/app/reader/utils.js]: Nit: This should be camel case, so getHearingWorksheetLink.

File: app/controllers/hearings_controller.rb [hunk 10]
```diff
@@ -29,7 +29,7 @@ def hearing_id
   end
 
   def verify_access
-    verify_authorized_roles("Hearing Prep")
+    verify_authorized_roles("Hearing Prep", "Reader")
```
Comment (sharonwarner) [inline app/controllers/hearings_controller.rb]: Let's update this to only allow Reader users access to the worksheet. Let me know if you want to pair on it.

**Comment by lakohl** [comment]
So the Hearing Worksheet link will show on the cases summary of all the appeal streams that appear on the Hearing Worksheet? (ideal) or just on the one that triggered the scheduling of the hearing?

**Comment by lakohl** [comment]
Better way to ask ^ is how are we showing multiple active appeal streams (not yet merged or unable to be merged) in Reader?

**Comment by sharonwarner** [comment]
@lakohl Just the appeal that triggered the hearing right now.

File: client/app/reader/utils.js (hunk 9 above)
Comment (nolawi) [reply client/app/reader/utils.js]: since we are using it as a component, has to start with capital letter

**Comment by gnakm** [comment]
@sharonwarner I'm AC creeping here but could we take out the download icon since it's now read-only?

**Comment by sharonwarner** [comment]
@gnakm Here's a screenshot of the page that we're sending users to, let me know if you still want me to remove the download link. 

screen shot 2017-12-19 at 6 19 13 pm

**Comment by gnakm** [comment]
Yes please. I think it makes sense to take it out.

File: client/app/components/RenderFunctions.jsx [hunk 11]
```diff
@@ -344,7 +344,7 @@ export class DownloadIcon extends React.PureComponent {
   render() {
     return <svg width="14px" height="19px" viewBox="0 0 14 19" version="1.1" xmlns="http://www.w3.org/2000/svg">
       <g stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
-        <g fill="#FFFFFF">
+        <g>
```
Comment (mdbenjam) [inline client/app/components/RenderFunctions.jsx]: Can we not change this file if we've removed the download icon?

File: app/models/user.rb [hunk 12]
```diff
@@ -101,7 +101,12 @@ def current_case_assignments_with_views
     opened_appeals = viewed_appeals(appeals.map(&:id))
 
     appeals.map do |appeal|
-      appeal.to_hash(viewed: opened_appeals[appeal.id], issues: appeal.issues)
+      hearings = Hearing.where(appeal_id: appeal.id)
+      appeal.to_hash(
+        viewed: opened_appeals[appeal.id],
+        issues: appeal.issues,
+        hearings: hearings.map(&:to_hash_for_reader)
```
Comment (mdbenjam) [inline app/models/user.rb]: This is going to break my current speed optimization :/

Comment (mdbenjam) [reply app/models/user.rb]: This should actually be pretty easy to move over into case_assignment.rb. Then it'll be fast :).

File: lib/fakes/appeal_repository.rb [hunk 13]
```diff
@@ -19,7 +19,11 @@ def clean!
     end
 
     def load_user_case_assignments_from_vacols(_css_id)
-      appeal_records || Fakes::Data::AppealData.default_records
+      user_case_assignments = appeal_records || Fakes::Data::AppealData.default_records
+      appeal = user_case_assignments.first
+      # Create a fake hearing for the first appeal if one doesn't already exist
+      Generators::Hearing.create(appeal: appeal) if Hearing.where(appeal: appeal).length == 0
```
Comment (mdbenjam) [inline lib/fakes/appeal_repository.rb]: Might be interesting to add two hearings since we can have more than one.

File: app/models/user.rb [hunk 14]
```diff
@@ -102,7 +102,12 @@ def current_case_assignments_with_views
     opened_appeals = viewed_appeals(appeals.map(&:id))
 
     appeals.map do |appeal|
-      appeal.to_hash(viewed: opened_appeals[appeal.id], issues: appeal.issues)
+      hearings = Hearing.where(appeal_id: appeal.id)
```
Comment (mdbenjam) [inline app/models/user.rb]: Can we do this like we did the above viewed_appeals? It's a lot more efficient to do it all in one query.

File: app/models/user.rb [hunk 15]
```diff
@@ -100,9 +100,14 @@ def station_offices
...
     appeals = current_case_assignments
     opened_appeals = viewed_appeals(appeals.map(&:id))
+    appeal_hearings = appeal_hearings(appeals.map(&:id))
 
     appeals.map do |appeal|
-      appeal.to_hash(viewed: opened_appeals[appeal.id], issues: appeal.issues)
+      appeal.to_hash(
+        viewed: opened_appeals[appeal.id],
+        issues: appeal.issues,
+        hearings: appeal_hearings.where(appeal: appeal)
```
Comment (mdbenjam) [inline app/models/user.rb]: I don't think you need this second `where` :)

**Pull request merged**